File: two_of_us/doctypes/challenge/challenge.js
// -*- coding: utf-8 -*-
/*
 * Challenge Doctype Client Script
 */

frappe.ui.form.on('Challenge', {
    refresh: function (frm) {
        // Add custom buttons
        if (frm.doc.status === 'active') {
            frm.add_custom_button(__('Complete Challenge'), function () {
                frm.set_value('status', 'completed');
                frm.save();
            });
        }

        if (frm.doc.status === 'upcoming') {
            frm.add_custom_button(__('Start Challenge'), function () {
                frm.set_value('status', 'active');
                frm.save();
            });
        }
    },

    start_date: function (frm) {
        // Validate start date
        if (frm.doc.start_date && frm.doc.end_date) {
            if (frm.doc.start_date > frm.doc.end_date) {
                frappe.msgprint(__('Start date cannot be after end date'));
                frm.set_value('start_date', '');
            }
        }
    },

    end_date: function (frm) {
        // Validate end date
        if (frm.doc.start_date && frm.doc.end_date) {
            if (frm.doc.start_date > frm.doc.end_date) {
                frappe.msgprint(__('End date cannot be before start date'));
                frm.set_value('end_date', '');
            }
        }
    },

    status: function (frm) {
        // Auto-set completion date when status changes to completed
        if (frm.doc.status === 'completed' && !frm.doc.completed_date) {
            frm.set_value('completed_date', frappe.datetime.get_today());
        }

        // Auto-set progress when status changes
        if (frm.doc.status === 'completed') {
            frm.set_value('progress', 100);
        } else if (frm.doc.status === 'active') {
            frm.set_value('progress', 0);
        }
    }
});

File: two_of_us/doctypes/challenge/challenge.py
import frappe
from frappe import _
from datetime import date


class Challenge(frappe.model.document.Document):
    def validate(self):
        """Validate challenge data"""
        if self.start_date and self.end_date and self.start_date > self.end_date:
            frappe.throw(_("Start date cannot be after end date"))

        if self.points < 0:
            frappe.throw(_("Points cannot be negative"))

    def on_update(self):
        """Handle challenge updates"""
        if self.status == "completed" and not hasattr(self, "completed_date"):
            self.completed_date = date.today()
            self.points_earned = self.points
            self.update_couple_profile()

    def update_couple_profile(self):
        """Update couple profile when challenge is completed"""
        if self.status == "completed":
            # Update couple profile statistics
            profile = frappe.get_doc(
                "Couple Profile",
                {
                    "partner1_name": self.partner1_name,
                    "partner2_name": self.partner2_name,
                },
            )

            if profile:
                profile.relationship_score += self.points
                profile.save()
